File: preview-server.mjs
import http from "node:http";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = __dirname;
const PORT = 4190;

const MIME = {
  ".html": "text/html; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".svg": "image/svg+xml",
  ".png": "image/png",
  ".txt": "text/plain; charset=utf-8",
  ".xml": "application/xml; charset=utf-8",
  ".ico": "image/x-icon",
};

function isBlocked(urlPath) {
  const lowered = urlPath.toLowerCase();
  if (lowered.includes("/.") || lowered.startsWith("/.")) return true;
  const blocked = [
    "/readme",
    "/readme.md",
    "/.git",
    "/.env",
    "/node_modules",
    "/preview-server.py",
    "/preview-server.mjs",
    "/.gitignore",
    "/build-render.mjs",
    "/render.yaml",
    "/package.json",
    "/package-lock.json",
  ];
  return blocked.some((item) => lowered === item || lowered.startsWith(item + "/"));
}

function send(res, status, type, body) {
  res.writeHead(status, {
    "Content-Type": type,
    "Content-Length": Buffer.byteLength(body),
    "Cache-Control": "no-store",
  });
  res.end(body);
}

function send404(res) {
  const page = path.join(ROOT, "404.html");
  const body = fs.existsSync(page) ? fs.readFileSync(page) : Buffer.from("Not found");
  send(res, 404, "text/html; charset=utf-8", body);
}

const server = http.createServer((req, res) => {
  try {
    const url = new URL(req.url || "/", `http://127.0.0.1:${PORT}`);
    let pathname = decodeURIComponent(url.pathname);
    if (pathname === "/" || pathname === "") pathname = "/index.html";
    if (isBlocked(pathname)) return send404(res);

    const candidate = path.resolve(ROOT, "." + pathname);
    if (!candidate.startsWith(ROOT)) return send404(res);
    if (!fs.existsSync(candidate) || !fs.statSync(candidate).isFile()) return send404(res);

    const ext = path.extname(candidate).toLowerCase();
    const type = MIME[ext] || "application/octet-stream";
    send(res, 200, type, fs.readFileSync(candidate));
  } catch {
    send404(res);
  }
});

server.listen(PORT, "127.0.0.1", () => {
  console.log(`Serving ${ROOT} on http://127.0.0.1:${PORT}`);
});

File: preview-server.py
import http.server
import os
import socketserver
from pathlib import Path
from urllib.parse import unquote, urlparse

ROOT = Path(__file__).resolve().parent
PORT = 4190

MIME = {
    ".html": "text/html; charset=utf-8",
    ".css": "text/css; charset=utf-8",
    ".js": "text/javascript; charset=utf-8",
    ".svg": "image/svg+xml",
    ".png": "image/png",
    ".txt": "text/plain; charset=utf-8",
    ".xml": "application/xml; charset=utf-8",
    ".ico": "image/x-icon",
}


class Handler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=str(ROOT), **kwargs)

    def do_GET(self):
        parsed = urlparse(self.path)
        path = unquote(parsed.path)
        if path in ("", "/"):
            path = "/index.html"

        # Deny sensitive paths
        lowered = path.lower()
        blocked_prefixes = (
            "/.",
            "/readme",
            "/readme.md",
            "/.git",
            "/.env",
            "/node_modules",
            "/preview-server",
            "/build-render.mjs",
            "/render.yaml",
            "/.gitignore",
            "/package.json",
            "/package-lock.json",
        )
        if any(lowered.startswith(p) or lowered == p.rstrip("/") for p in blocked_prefixes) or "/." in lowered:
            return self.send_branded_404()

        candidate = (ROOT / path.lstrip("/")).resolve()
        try:
            candidate.relative_to(ROOT)
        except ValueError:
            return self.send_branded_404()

        if candidate.is_file():
            return self.serve_file(candidate)

        return self.send_branded_404()

    def serve_file(self, path: Path):
        data = path.read_bytes()
        ctype = MIME.get(path.suffix.lower(), "application/octet-stream")
        self.send_response(200)
        self.send_header("Content-Type", ctype)
        self.send_header("Content-Length", str(len(data)))
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(data)

    def send_branded_404(self):
        page = ROOT / "404.html"
        data = page.read_bytes() if page.exists() else b"Not found"
        self.send_response(404)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.send_header("Content-Length", str(len(data)))
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(data)

    def log_message(self, fmt, *args):
        sys_stderr = __import__("sys").stderr
        sys_stderr.write("%s - %s\n" % (self.address_string(), fmt % args))


if __name__ == "__main__":
    os.chdir(ROOT)
    with socketserver.TCPServer(("127.0.0.1", PORT), Handler) as httpd:
        print(f"Serving {ROOT} on http://127.0.0.1:{PORT}")
        httpd.serve_forever()
